### PR TITLE
initial signature

### PR DESCRIPTION
**Pull request by Andries-Smit**

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+- sudo apt-get install -y --allow-downgrades subversion=1.7.22-1+WANdisco libsvn1=1.7.22-1+WANdisco
+- svn --version
+after_success:
+- bash <(curl -s https://codecov.io/bash)
+cache:
+  directories:
+  - node_modules
+env:
+  global:
+  - MX_APP_NAME=enumswitch-main
```
Comment (Andries-Smit) [inline .travis.yml]: dont think this is right

File: Gruntfile.js [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+const webpack = require("webpack");
+const webpackConfig = require("./webpack.config");
+const merge = require("webpack-merge");
+
+const webpackConfigRelease = webpackConfig.map(config => merge(config, {
+    devtool: false,
+    optimization: {
+        minimize: true
+    }
+}));
```
Comment (Andries-Smit) [inline Gruntfile.js:9]: this does not like the be webpack 4 compatible
requires `mode: production`

File: LICENSE [hunk 3]
```diff
@@ -186,7 +186,7 @@
       same "printed page" as the copyright notice for easier
       identification within third-party archives.
 
-   Copyright [yyyy] [name of copyright owner]
+   Copyright 2018 flock of birds.
```
Comment (Andries-Smit) [inline LICENSE]: 2018 Mendix

File: package.json [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+    "deployApp": "node ./node_modules/mendix-widget-build-script/dist/DeployMxApp.js",
+    "deployAndTest": "npm run deploy && npm run test:e2e"
+  },
+  "pre-commit": [
+    "emit",
+    "lint"
+  ],
+  "repository": {
+    "type": "git",
+    "url": "https://github.com/SamuelMuloki/Signature.git"
```
Comment (Andries-Smit) [inline package.json]: change to mendixlabs

File: src/Signature.webmodeler.ts [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+export class preview extends Component<SignatureContainerProps> {
+    render() {
+        return createElement(Signature, this.transformProps(this.props));
+    }
+
+    private transformProps(props: SignatureContainerProps): SignatureProps {
+        return {
+            height: props.height,
+            width: props.width,
+            gridx: props.gridx,
```
Comment (Andries-Smit) [inline src/Signature.webmodeler.ts]: hight widget props requires als % etc not always px

File: src/Signature.webmodeler.ts [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+            maxLineWidth: props.maxLineWidth,
+            minLineWidth: props.minLineWidth,
+            velocityFilterWeight: props.velocityFilterWeight,
+            showGrid: props.showGrid,
+            timeout: props.timeout
+        };
+    }
+}
+
+export function getPreviewCss() { return require("./ui/Signature.scss"); }
```
Comment (Andries-Smit) [inline src/Signature.webmodeler.ts]: make 3 lines

File: src/Signature.xml [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+<widget id="com.mendix.widget.custom.signature.Signature" offlineCapable="true" needsEntityContext="true" mobile="true" xmlns="http://www.mendix.com/widget/1.0/">
+    <name>Signature</name>
+    <description>A signature pad for capturing signatures</description>
+    <icon>iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAAehJREFUeNpi/J82k4Hh/38Ghn//UPAD5e0M99V2MUxeHcAAAlM2bLD5w8BQ9I6B4eaT//8rGaCAhQENPFDZwXAfqBkZ/P//Xy3GwcF4r4lJIOOlS6/u79y5Fyi8B8WAy7xbGS7zbGWQVGbE0Pzx40fPPXv2TFi3fv21f93d739DNYMA46VCX4bL3JvhGiS1GJE0M9h//cpQqqoT6P30ycOPvvHXZhT/+FGhAvIyFDAha0a1GaLZ3a/K++/vf69PHT17WVr0d8U1RlQXMhHSfPvm1deb1m24+es3g62sxD8MtSzEav4H1XsdqGAdsgFff0CdAnQZCzOD3IOrPHaByZHee3ftfHnh7JnbMM0Hz4AwFhfADGBmYmD8+48xpaBpf21xhunC778YbgmJMLYdufAfrBkn8DCD4LZUxp5nV5f9zwliXO5nzehpos7IQAqQAnk/w59xlZQIoxeQrQrECkAsDcRiQCwIxDxAzA4NN7DpvHL6QJ02cWA2KAkAsTgQiwKxMBALADEfEHMDMScQs0E1g2KOEaTZuHw3A7NaZA8Dn7zBL3Z+iS9A9ldGRsZvQMnvQPYPIPsnkP0byP4DZP8Fsv8B2f+BbAYFr1JISgSaEgWkl0G9QjKbiYFCAHKBCzSgyAGvAAIMAHD/1dmmASuxAAAAAElFTkSuQmCC</icon>
+
+    <properties>
+        <property key="dataUrl" type="attribute" required="false" isPath="optional" pathType="reference">
```
Comment (Andries-Smit) [inline src/Signature.xml]: remove prop

File: src/Signature.xml [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+            <caption>Minimum line width</caption>
+            <category>Pen Customization</category>
+            <description>Pathline minimum width</description>
+        </property>
+        <property key="velocityFilterWeight" type="string" defaultValue="0.7">
+            <caption>Pen pressure weight</caption>
+            <category>Pen Customization</category>
+            <description>Velocity filter pressure</description>
+        </property>
+        <property key="editable" type="enumeration" defaultValue="default">
```
Comment (Andries-Smit) [inline src/Signature.xml]: Remove

File: src/Signature.xml [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+            <category>Appearance</category>
+            <description>Border width(px)</description>
+        </property>
+        <property key="saveImage" type="enumeration" defaultValue="onChange">
+            <caption>Save image</caption>
+            <category>Events</category>
+            <description>Save signature image, OnChange is default where the image is saved based on a given timeout</description>
+            <enumerationValues>
+                <enumerationValue key="onChange">OnChange</enumerationValue>
+                <enumerationValue key="onFormCommit">OnFormCommit</enumerationValue>
```
Comment (Andries-Smit) [inline src/Signature.xml]: One capital, not camel cased

File: src/Signature.xml [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+            <caption>Grid color</caption>
+            <category>Appearance</category>
+            <description>Grid color</description>
+        </property>
+        <property key="gridBorder" type="integer" defaultValue="1">
+            <caption>Grid border width</caption>
+            <category>Appearance</category>
+            <description>Border width(px)</description>
+        </property>
+        <property key="saveImage" type="enumeration" defaultValue="onChange">
```
Comment (Andries-Smit) [inline src/Signature.xml]: `onFormCommit` should be default

File: src/Signature.xml (hunk 9 above)
Comment (Andries-Smit) [inline src/Signature.xml]: order values, default first

File: src/Signature.xml [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+        </property>
+        <property key="timeout" type="integer" defaultValue="1000">
+            <caption>Signature timeout</caption>
+            <category>Events</category>
+            <description>Time (in ms) to wait before considering the signature finished after the pen has stopped touching the pad</description>
+        </property>
+        <property key="afterSignEvent" type="enumeration" defaultValue="doNothing">
+            <caption>After sign</caption>
+            <category>Events</category>
+            <description/>
```
Comment (Andries-Smit) [inline src/Signature.xml]: After sign microflow will be triggerd on the timeing.... right? need to be documented

File: src/components/Signature.ts [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+        this.drawGrid = this.drawGrid.bind(this);
+    }
+
+    render() {
+        return createElement("div", {
+            className: classNames("widget-signature-wrapper", { disabled: this.props.status === "disabled" }),
+            style: this.getStyle(this.props)
+        }, createElement("canvas", {
+                className: classNames("widget-Signature", "form-control mx-textarea-input mx-textarea",
+                { disabled: this.props.status === "disabled" }),
```
Comment (Andries-Smit) [inline src/components/Signature.ts]: indenting wrong

File: src/components/Signature.ts [hunk 13]
```diff
@@ -0,0 +1,175 @@
...
+                minWidth: this.props.minLineWidth,
+                onEnd: this.handleSignEnd
+            });
+
+            if (this.canvasNode.parentElement) {
+                this.height = this.canvasNode.parentElement.clientHeight;
+                this.width = this.canvasNode.parentElement.clientWidth;
+            }
+        }
+        window.addEventListener("resize", this.resizeCanvas);
```
Comment (Andries-Smit) [inline src/components/Signature.ts]: tricky to connect to resize event on the window. as in the webmodelere you are working in a iframe..

File: src/components/SignatureContainer.ts [hunk 14]
```diff
@@ -0,0 +1,251 @@
...
+    url: string;
+    alertMessage: string;
+    base64Uri: string;
+    clearPad: boolean;
+}
+
+export default class SignatureContainer extends Component<SignatureContainerProps, SignatureContainerState> {
+    private subscriptionHandles: number[] = [];
+    private _formCommitHandle = 0;
+    public get formHandle() {
```
Comment (Andries-Smit) [inline src/components/SignatureContainer.ts]: whats happening here?

File: src/components/SignatureContainer.ts [hunk 15]
```diff
@@ -0,0 +1,251 @@
...
+interface SignatureContainerState {
+    url: string;
+    alertMessage: string;
+    base64Uri: string;
+    clearPad: boolean;
+}
+
+export default class SignatureContainer extends Component<SignatureContainerProps, SignatureContainerState> {
+    private subscriptionHandles: number[] = [];
+    private _formCommitHandle = 0;
```
Comment (Andries-Smit) [inline src/components/SignatureContainer.ts]: private variables are should not be prefixed

File: src/components/SignatureContainer.ts [hunk 16]
```diff
@@ -0,0 +1,251 @@
...
+    public get formHandle() {
+        return this._formCommitHandle;
+    }
+    public set formHandle(value) { this._formCommitHandle = value; }
+
+    constructor(props: SignatureContainerProps) {
+        super(props);
+
+        this.state = {
```
Comment (Andries-Smit) [inline src/components/SignatureContainer.ts]: move state init to declaration new react way

File: src/Signature.webmodeler.ts [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+
+export function getVisibleProperties(valueMap: SignatureContainerProps, visibilityMap: VisibilityMap) {
+    visibilityMap.afterSignMicroflow = valueMap.afterSignEvent === "callMicroflow";
+    visibilityMap.afterSignNanoflow = valueMap.afterSignEvent === "callNanoflow";
```
Comment (Andries-Smit) [inline src/Signature.webmodeler.ts]: only who `timeout` on `Save image on change timeout`

File: src/Signature.xml [hunk 18]
```diff
@@ -0,0 +1,136 @@
...
+        <property key="saveImage" type="enumeration" defaultValue="onChange">
+            <caption>Save image</caption>
+            <category>Events</category>
+            <description>Save signature image, OnChange is default where the image is saved based on a given timeout</description>
+            <enumerationValues>
+                <enumerationValue key="onChange">OnChange</enumerationValue>
+                <enumerationValue key="onFormCommit">OnFormCommit</enumerationValue>
+            </enumerationValues>
+        </property>
+        <property key="timeout" type="integer" defaultValue="1000">
```
Comment (Andries-Smit) [inline src/Signature.xml]: rename key to `changeTimeout`

File: src/components/Signature.ts (hunk 13 above)
Comment (Andries-Smit) [reply src/components/Signature.ts]: Resizing should ALWAYS be throttled, else every pixel change will cause a redraw. should only happen every 50 miliseconds

**Review by Andries-Smit — COMMENTED** [review]
Not all is reviewed, more comment on the functional implementation is shared on the chat

File: .travis.yml (hunk 1 above)
Comment (SamuelMuloki) [reply .travis.yml]: done

File: Gruntfile.js (hunk 2 above)
Comment (SamuelMuloki) [reply Gruntfile.js:9]: done

File: LICENSE (hunk 3 above)
Comment (SamuelMuloki) [reply LICENSE]: done

File: package.json (hunk 4 above)
Comment (SamuelMuloki) [reply package.json]: done

File: src/Signature.webmodeler.ts (hunk 6 above)
Comment (SamuelMuloki) [reply src/Signature.webmodeler.ts]: done

File: src/Signature.xml (hunk 9 above)
Comment (SamuelMuloki) [reply src/Signature.xml]: done

File: src/Signature.xml (hunk 10 above)
Comment (SamuelMuloki) [reply src/Signature.xml]: done

File: src/Signature.xml (hunk 9 above)
Comment (SamuelMuloki) [reply src/Signature.xml]: done

File: src/components/Signature.ts (hunk 12 above)
Comment (SamuelMuloki) [reply src/components/Signature.ts]: done

File: src/Signature.xml (hunk 8 above)
Comment (SamuelMuloki) [reply src/Signature.xml]: done

File: src/Signature.xml (hunk 7 above)
Comment (SamuelMuloki) [reply src/Signature.xml]: done

File: src/components/SignatureContainer.ts (hunk 14 above)
Comment (SamuelMuloki) [reply src/components/SignatureContainer.ts]: removed

File: src/Signature.xml (hunk 18 above)
Comment (SamuelMuloki) [reply src/Signature.xml]: done

File: src/components/SignatureContainer.ts (hunk 16 above)
Comment (SamuelMuloki) [reply src/components/SignatureContainer.ts]: done

**Pull request merged**